### ISSUE TITLE
Testers for release 2.3.0 are welcome

### ISSUE THREAD

**Issue by TomFreudenberg** [4commerce-technologies-AG/midi-smtp-server#15]
Hello MidiSmtpServer community,

we made a number of addtions and fixes to the new release 2.3.0.

1. Support [IPv4 and IPv6 (documentation)](https://github.com/4commerce-technologies-AG/midi-smtp-server#ipv4-and-ipv6-ready)
2. Support binding of [multiple ports and hosts / ip addresses](https://github.com/4commerce-technologies-AG/midi-smtp-server#multiple-ports-and-addresses)
3. Support (optionally) SMTP [PIPELINING](https://tools.ietf.org/html/rfc2920) extension
4. Support (optionally) SMTP [8BITMIME](https://github.com/4commerce-technologies-AG/midi-smtp-server#8bitmime-and-smtputf8-support) extension
5. Support (optionally) SMTP [SMTPUTF8](https://github.com/4commerce-technologies-AG/midi-smtp-server#8bitmime-and-smtputf8-support) extension
6. The SMTP PIPELINING, 8BITMIME and SMTPUTF8 extensions are disabled by _default_.

Before publishing a new release, it would be great if some of you could check the new component with their existing projects.

There should be NO COMPATBILITY ISSUES - so its fine, when all will run as it already run.

Maybe the new possibilities are interesting to you, than it would be great if you spend some time on them too.

Thanks for your feedback.
Tom

**Comment by TomFreudenberg** [comment]
Hi @gencer, @mwhobrey, @kaspergrubbe, @chrise86, yms9654 would be great if you may test the new release.

Just use the **lastest master branch**

Cheers
Tom

**Comment by gencer** [comment]
I'm on it today. Will report back.

**Comment by TomFreudenberg** [comment]
Hi @gencer - do you have an results?

**Comment by TomFreudenberg** [comment]
Hi @mwhobrey may I ask you for some short "live" test results?

**Comment by gencer** [comment]
@TomFreudenberg, I've successfully tested master branch except with TLS. I will test TLS in few days. AUTH and NONAUTH is success. However one thing makes me wonder. When I send e-mail from gmail to my server i got some errors. In first 2 tests it was okay but in third and fourth I got some interesting debug logs.

```
2018-10-08 07:23:01 +0000: [DEBUG] >>> 250 Requested mail action okay, completed
2018-10-08 07:23:01 +0000: [DEBUG] <<< QUIT
2018-10-08 07:23:01 +0000: [DEBUG] Client disconnect from 209.85.217.41:44512 on 10.0.1.4:25
2018-10-08 07:28:15 +0000: [DEBUG] Client connect from 13.90.139.151:38458 to 10.0.1.4:25
2018-10-08 07:28:15 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:28:15 +0000: [DEBUG] Client connect from 52.179.17.245:43072 to 10.0.1.4:25
2018-10-08 07:28:15 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:28:16 +0000: [DEBUG] Client connect from 13.84.161.169:59224 to 10.0.1.4:25
2018-10-08 07:28:16 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:28:16 +0000: [ERROR] Connection reset by peer
2018-10-08 07:28:16 +0000: [DEBUG] IOError - Can't send 421 abort code!
2018-10-08 07:28:16 +0000: [DEBUG] Client disconnect from 13.84.161.169:59224 on 10.0.1.4:25
2018-10-08 07:28:20 +0000: [ERROR] Connection reset by peer
2018-10-08 07:28:20 +0000: [DEBUG] IOError - Can't send 421 abort code!
2018-10-08 07:28:20 +0000: [DEBUG] Client disconnect from 13.90.139.151:38458 on 10.0.1.4:25
2018-10-08 07:28:26 +0000: [DEBUG] Client connect from 52.179.17.245:43073 to 10.0.1.4:25
2018-10-08 07:28:26 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:28:37 +0000: [DEBUG] Client connect from 13.84.161.169:57696 to 10.0.1.4:25
2018-10-08 07:28:37 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:28:37 +0000: [DEBUG] Client connect from 13.90.139.151:36856 to 10.0.1.4:25
2018-10-08 07:28:37 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:28:37 +0000: [ERROR] Connection reset by peer
2018-10-08 07:28:37 +0000: [DEBUG] IOError - Can't send 421 abort code!
2018-10-08 07:28:37 +0000: [DEBUG] Client disconnect from 13.90.139.151:36856 on 10.0.1.4:25
2018-10-08 07:28:42 +0000: [ERROR] Connection reset by peer
2018-10-08 07:28:42 +0000: [DEBUG] IOError - Can't send 421 abort code!
2018-10-08 07:28:42 +0000: [DEBUG] Client disconnect from 13.84.161.169:57696 on 10.0.1.4:25
2018-10-08 07:28:45 +0000: [DEBUG] Client connect from 13.84.161.169:59120 to 10.0.1.4:25
2018-10-08 07:28:45 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:28:46 +0000: [ERROR] MidiSmtpServer::SmtpdIOTimeoutException
2018-10-08 07:28:46 +0000: [DEBUG] Client disconnect from 52.179.17.245:43072 on 10.0.1.4:25
2018-10-08 07:28:56 +0000: [ERROR] Connection reset by peer
2018-10-08 07:28:56 +0000: [DEBUG] IOError - Can't send 421 abort code!
2018-10-08 07:28:56 +0000: [DEBUG] Client disconnect from 52.179.17.245:43073 on 10.0.1.4:25
2018-10-08 07:29:01 +0000: [DEBUG] Client connect from 209.85.222.52:34704 to 10.0.1.4:25
2018-10-08 07:29:01 +0000: [DEBUG] >>> 220 10.0.1.4 says welcome!
2018-10-08 07:29:01 +0000: [DEBUG] <<< EHLO mail-ua1-f52.google.com
2018-10-08 07:29:01 +0000: [DEBUG] >>> 250-10.0.1.4 at your service!
250-AUTH LOGIN PLAIN
250 OK
2018-10-08 07:29:01 +0000: [DEBUG] <<< MAIL FROM:<XXX@gmail.com>
2018-10-08 07:29:01 +0000: [DEBUG] >>> 250 OK
2018-10-08 07:29:01 +0000: [DEBUG] <<< RCPT TO:<XXX@mydomain.com>
2018-10-08 07:29:01 +0000: [DEBUG] >>> 250 OK
2018-10-08 07:29:02 +0000: [DEBUG] <<< DATA
2018-10-08 07:29:02 +0000: [DEBUG] >>> 354 Enter message, ending with "." on a line by itself
2018-10-08 07:29:02 +0000: [DEBUG] mail reveived at: [10.0.1.4:25] from: [<XXX@gmail.com>] for recipient(s): [["<XXX@mydomain.com>"]]...
2018-10-08 07:29:02 +0000: [DEBUG] Received: by mail-ua1-f52.google.com with SMTP id 
--0000000000009cdb3e0577b28f95--
2018-10-08 07:29:02 +0000: [DEBUG] >>> 250 Requested mail action okay, completed
2018-10-08 07:29:02 +0000: [DEBUG] <<< QUIT
2018-10-08 07:29:02 +0000: [DEBUG] Client disconnect from 209.85.222.52:34704 on 10.0.1.4:25
2018-10-08 07:29:03 +0000: [ERROR] Connection reset by peer
2018-10-08 07:29:03 +0000: [DEBUG] IOError - Can't send 421 abort code!
2018-10-08 07:29:03 +0000: [DEBUG] Client disconnect from 13.84.161.169:59120 on 10.0.1.4:25
```

Is that a normal behavior?

**Comment by TomFreudenberg** [comment]
Hi @gencer 

thanks for feedback.

I guess, you are spotting on the IOError messages?

<pre>
2018-10-08 07:28:56 +0000: [ERROR] Connection reset by peer
2018-10-08 07:28:56 +0000: [DEBUG] IOError - Can't send 421 abort code!
2018-10-08 07:28:56 +0000: [DEBUG] Client disconnect from 52.179.17.245:43073 on 10.0.1.4:25
</pre>

I have seen this a number of times. Normally this comes up in log, when someone or you monitoring the service in just opening and closing a TCP channel. This is done by them like

``` ruby
socket.connect
socket.close
```

If this comes up, the library still can't even send welcome message etc. to this clients. Due to this error condition, the library would like to signal 421 as an overall error, but also this can't be sent due to `io` lost signal.

So, yes this is normal behavior. I asked myself a number of times to log or not to log those rare conditions. But I had decided to show them in debugging log and so we will get informed about any connection.

**Comment by mwhobrey** [comment]
Hey @TomFreudenberg,

Sorry for the delay, I was out of town for the weekend. I will hopefully have some time to run this through some tests this week if I can keep all the fires out that sprung up while I was away this weekend.

**Comment by TomFreudenberg** [comment]
Hi Michael @mwhobrey 

thanks for feedback and sorry for needling you. Hopefully nothing will be changed when you update with master branch - its what I expect.

Cheers
Tom

**Comment by mwhobrey** [comment]
Hey @TomFreudenberg,

No worries! I'm happy to help out in any way I can. 

Thanks for all the work!

**Comment by TomFreudenberg** [comment]
@gencer and @mwhobrey 

I made an update on library to enable unit testing for methods _process_line_ etc.

https://github.com/4commerce-technologies-AG/midi-smtp-server/commit/ad5bc2cfa14c398d24de42bb84e4e91b51760485

For that I changed the explicit call of `Thread.current` into a _session_ argument to that methods.

It should not break anything but maybe you may update for latest branch and have a look on that too.

Thanks for help
Tom

P.S.: I am currently implementing some integration tests to have valid checks on upcoming changes.

**Comment by mwhobrey** [comment]
Hey @TomFreudenberg,

I'm sorry for the delay in testing this. My dev laptop went out on me a few weeks ago and I have been configuring and setting up things as needed. I **think** I have just about gotten everything where it needs to be on my new laptop to run through my tests successfully. 

I have already gotten several partial successes, the only reason they are failing now is setup issues I forgot about on my end, so this looks very good so far! This is #1 on my queue for today so I will keep you posted.

Thanks for the hard work,

Mike

**Comment by mwhobrey** [comment]
Hey @TomFreudenberg,

I updated to a9c8f05c60270264fca856c88f3a60bf4e7f924e and tested with my integrating apps and everything worked beautifully. I also pulled the code into my fork and ran the tests you have on windows and on ubuntu via WSL and all the tests passed. 

Let me know if there is anything specific I can test for you. I'll have my QA team do some more in depth testing over the next few days.

Thanks again!
Mike

**Comment by TomFreudenberg** [comment]
Hey Mike @mwhobrey 

thanks a lot for doing the work and given feedback. For future hopefully we got a number of representive testings but for 2.3.0 the real life feedback was important to me.

So I guess we are close to publish 2.3.0 as gem.

I am currently finishing the CRLF feature as last feature and addon to 2.3.0.

It is great to have you on board here.

Have a nice day
Tom

**Comment by TomFreudenberg** [comment]
Hi guys, thanks again for helping on this stuff.

I will append some more unit and integration tests but at least we should have by now a reliable update.

Close here, feel free to append on any occurence.

Cheers
Tom

**Comment by TomFreudenberg** [comment]
Hello guys

the current commit is now ready for release. If possible just check again you application and be aware of the new handling of max_connections which will drop now more than allowed concurrent connections.

Thanks for your help
Tom

P.S.: If there a no issues, I will release on Wednesday the 2.3.0

P.P.S.: Your help in writing tests (pull request) is also welcome. I am currently working on some integration tests but we might need / have a lot more of the specs and unit tests as well.

**Comment by mwhobrey** [comment]
Good morning @TomFreudenberg,

I have updated to 6b15846b15c97a2f021a73712dc250a7281ed67a and all seems well so far from an integration standpoint (i.e. my desktop applications that use my implementation are working beautifully). 

I did attempt to run through the tests again, and there were a bunch that failed this time around, all of which fail with the error 

> Minitest::UnexpectedError: NameError: uninitialized constant MidiSmtpServer::DEFAULT_SMTPD_MAX_CONNECTIONS

Being the server works great on our staging servers and only the tests fail, I am assuming that the test scripts were not updated to reflect the most recent changes, and am not concerned, but just wanted to give you a heads up (I always tell my guys "if you see something, say something"). 

Let me know if there is anything else I can do to help.

Thank you!
Mike
